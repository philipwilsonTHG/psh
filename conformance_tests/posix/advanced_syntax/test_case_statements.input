# Test advanced case statement functionality
# PSH supports comprehensive case statements with pattern matching (v0.17.0 feature)

echo "Testing advanced case statement functionality:"

# Test basic case statement
value="apple"
echo "Testing basic case with value: $value"
case "$value" in
    apple)
        echo "Found apple"
        ;;
    banana)
        echo "Found banana"
        ;;
    *)
        echo "Unknown fruit"
        ;;
esac

# Test multiple patterns
echo "Testing multiple patterns:"
value="dog"
case "$value" in
    cat|dog|bird)
        echo "Found animal: $value"
        ;;
    car|bike|plane)
        echo "Found vehicle: $value"
        ;;
    *)
        echo "Unknown category"
        ;;
esac

# Test wildcard patterns
echo "Testing wildcard patterns:"
filename="test.txt"
case "$filename" in
    *.txt)
        echo "Text file: $filename"
        ;;
    *.log)
        echo "Log file: $filename"
        ;;
    *.*)
        echo "File with extension: $filename"
        ;;
    *)
        echo "No extension: $filename"
        ;;
esac

# Test character class patterns
echo "Testing character class patterns:"
input="A"
case "$input" in
    [A-Z])
        echo "Uppercase letter: $input"
        ;;
    [a-z])
        echo "Lowercase letter: $input"
        ;;
    [0-9])
        echo "Digit: $input"
        ;;
    *)
        echo "Other character: $input"
        ;;
esac

# Test question mark patterns
echo "Testing question mark patterns:"
code="A1"
case "$code" in
    ??)
        echo "Two character code: $code"
        ;;
    ???)
        echo "Three character code: $code"
        ;;
    *)
        echo "Other length code: $code"
        ;;
esac

# Test complex patterns
echo "Testing complex patterns:"
item="file123.txt"
case "$item" in
    file[0-9]*.txt)
        echo "Numbered text file: $item"
        ;;
    file*.log)
        echo "File log: $item"
        ;;
    test*.*)
        echo "Test file: $item"
        ;;
    *)
        echo "Other file: $item"
        ;;
esac

# Test case statement with variables
echo "Testing case with variable patterns:"
pattern="*.txt"
file="document.txt"
case "$file" in
    $pattern)
        echo "File matches pattern: $file"
        ;;
    *)
        echo "File doesn't match: $file"
        ;;
esac

# Test case statement with command substitution
echo "Testing case with command substitution:"
current_user=$(whoami)
case "$current_user" in
    root)
        echo "Root user detected"
        ;;
    $USER)
        echo "Current user: $current_user"
        ;;
    *)
        echo "Other user: $current_user"
        ;;
esac

# Test fallthrough behavior with ;;&
echo "Testing fallthrough patterns:"
number="3"
case "$number" in
    [1-5])
        echo "Number 1-5: $number"
        ;;
    [2-8])
        echo "Number 2-8: $number"
        ;;
    *)
        echo "Other number: $number"
        ;;
esac

# Test case with nested structures
echo "Testing nested structures in case:"
type="file"
name="test.txt"
case "$type" in
    file)
        echo "Processing file: $name"
        case "$name" in
            *.txt)
                echo "  Text file processing"
                ;;
            *.log)
                echo "  Log file processing"
                ;;
            *)
                echo "  Unknown file type"
                ;;
        esac
        ;;
    directory)
        echo "Processing directory: $name"
        ;;
    *)
        echo "Unknown type: $type"
        ;;
esac

# Test case with function calls
check_file_type() {
    case "$1" in
        *.txt) echo "text" ;;
        *.log) echo "log" ;;
        *.tmp) echo "temporary" ;;
        *) echo "unknown" ;;
    esac
}

echo "Testing case in functions:"
echo "File type: $(check_file_type "data.txt")"
echo "File type: $(check_file_type "error.log")"
echo "File type: $(check_file_type "cache.tmp")"
echo "File type: $(check_file_type "readme")"

# Test case with arithmetic results
echo "Testing case with arithmetic:"
value=$((2 + 3))
case "$value" in
    [1-3])
        echo "Small number: $value"
        ;;
    [4-6])
        echo "Medium number: $value"
        ;;
    [7-9])
        echo "Large number: $value"
        ;;
    *)
        echo "Very large number: $value"
        ;;
esac

# Test case with empty patterns
echo "Testing edge cases:"
empty=""
case "$empty" in
    "")
        echo "Empty string matched"
        ;;
    *)
        echo "Non-empty string"
        ;;
esac

# Test case with special characters
echo "Testing special characters:"
special="[test]"
case "$special" in
    \[*\])
        echo "Bracketed text: $special"
        ;;
    \(*\))
        echo "Parenthesized text: $special"
        ;;
    *)
        echo "Other format: $special"
        ;;
esac

echo "Advanced case statement test completed"