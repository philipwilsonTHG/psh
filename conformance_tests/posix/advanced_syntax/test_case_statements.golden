Testing advanced case statement functionality:
Testing basic case with value: apple
Found apple
Testing multiple patterns:
Found animal: dog
Testing wildcard patterns:
Text file: test.txt
Testing character class patterns:
Uppercase letter: A
Testing question mark patterns:
Two character code: A1
Testing complex patterns:
Numbered text file: file123.txt
Testing case with variable patterns:
File matches pattern: document.txt
Testing case with command substitution:
Current user: pwilson
Testing fallthrough patterns:
Number 1-5: 3
Testing nested structures in case:
Processing file: test.txt
  Text file processing
Testing case in functions:
File type: text
File type: log
File type: temporary
File type: unknown
Testing case with arithmetic:
Medium number: 5
Testing edge cases:
Empty string matched
Testing special characters:
Other format: [test]
Advanced case statement test completed
